aten/src/ATen/core/ivalue.cpp

```diff
@@ -1009,21 +1011,23 @@ TORCH_API intrusive_ptr<ivalue::Future> collectAny(
     return res;
   }
   TypePtr typePtr = srcs.get(0)->elementType();
+  const std::vector<c10::Device>& devices = srcs.get(0)->devices();
   for (const auto i : c10::irange(srcs.size())) {
     if (srcs.get(i)->completed()) {
       return srcs.get(i);
     }
     TORCH_CHECK(i == 0 || (*typePtr == *srcs.get(i)->elementType()));
+    TORCH_CHECK(i == 0 || (devices == srcs.get(i)->devices()));
```

Comment:
This is prior to this PR: `TORCH_CHECK` are usually user-facing checks with clear error messages. If this is internal only, maybe use `TORCH_INTERNAL_ASSERT` instead?

Reply:
This requires all Futures' value using the same set of devices? I guess this might be fine as we haven't exposed this to Python land. I am not sure if there are existing C++ use-cases of this API. cc @pritamdamania87

Reply:
> TORCH_CHECK are usually user-facing checks

I think this might be user-facing? I mean, the error happens if the user passes in bad values, which could totally happen. I don't know if the end user could call this directly, or if it's private, but within this function this is just an argument validation error. I can put a clearer error message.

> This requires all Futures' value using the same set of devices?

Alternatively we could make this function compute the union of the set of devices of all of its arguments and use that. I went for this other way because it was simpler and I expected it to be enough in practice (e.g., all futures obtained from RPC have the same set of devices). But it'll be very easy to relax this restriction if we'll need to.

Reply:
> I am not sure if there are existing C++ use-cases of this API. cc @pritamdamania87

Not aware of any usage of this API yet